Add UnusedAbstractValue and improve LiveVarAnalysis (#203)
eisop/checker-framework: dataflow/manual/content.tex

class NullnessValue extends CFAbstractValue<NullnessValue>
\end{verbatim}

\subsubsection{UnusedAbstractValue}
\label{sec:unused_abstract_value_class}

> UnusedAbstractValue, as the name suggests, is an AbstractValue but is

Reviewer: ```suggestion
UnusedAbstractValue, as the name suggests, is an AbstractValue that is
```

not involved in any computation during dataflow analysis. This class should
only be used as a type argument in transfer functions. It is implemented because some

> analyses don't need and don't have their own AbstractValue classes.

Reviewer: ```suggestion
analyses do not need and do not have their own AbstractValue classes.
```

> For example, in LiveVariable analysis (\autoref{sec:live_variable})
> , we need to compute the least upper bound of successors' stores, and

Reviewer: ```suggestion
For example, in LiveVariable analysis (\autoref{sec:live_variable}),
we need to compute the least upper bound of successors' stores, and
```

> it is meaningless to further calculate the least upper bound of two individual live variables in a smaller granularity.

Reviewer: ```suggestion
it is meaningless to further calculate the least upper bound of two
individual live variables at a smaller granularity.
```

> Since there's no computation between two AbstractValues, using the UnusedAbstractValue

Reviewer: ```suggestion
Since there is no computation between two AbstractValues, using the UnusedAbstractValue
```

class NullnessValue extends CFAbstractValue<NullnessValue>
\end{verbatim}

\subsubsection{UnusedAbstractValue}

> \label{sec:UnusedAbstractValue_classes}

Reviewer: ```suggestion
\label{sec:unused_abstract_value_class}
```

UnusedAbstractValue, as the name suggests, is an AbstractValue but is
not involved in any computation during dataflow analysis. That is, the
UnusedAbstractValue implements the leastUpperBound method in a way it

> just throws an error if it is called, without computing the lub with the

Reviewer: The rest of this manual doesn't use `lub`. Please expand.

Author: done.

class NullnessValue extends CFAbstractValue<NullnessValue>
\end{verbatim}

> \subsubsection{UnusedAbsValue}

Author: fixed it.

\label{sec:UnusedAbstractValue_classes}

> UnusedAbstractValue, as the name suggests, is an AbstractValue but is

Reviewer 2: Maybe `NoMergeAbstractValue` since it sounds more descriptive to me.

Author: Sure, I think that's a descriptive name, I will talk with Prof about it on Monday

Reviewer: This class should only be used as a type argument in transfer functions. In this first paragraph, the important point isn't what LUB computes, it is why we have this class and when to use it.

Author: added.

not involved in any computation during dataflow analysis. That is, the
UnusedAbstractValue implements the leastUpperBound method in a way it
just throws an error if it is called, without computing the lub with the
other AbstractValue.

> It can be used in an analysis which does not require any computation

Reviewer 2: You can mention that, in some situations, lub can be computed on the store level, followed by the example you have.

Author: Done.

between two AbstractValues. For example, in LiveVariable analysis (\autoref{sec:live_variable}), we
need to compute the lub of successors' stores, and it is meaningless to

> further calculate the lub of several live variables in a smaller granularity.

Reviewer 2: ```suggestion
further calculate the lub of two individual live variables in a smaller granularity.
```
